PubSub: Subscriber not pulling from pubsub queue

#### Environment details

OS type and version: GKE node n2d-highmem-8
Java version: 21
version(s): google-cloud-pubsub-v1-1.120.0

#### Steps to reproduce

  1. Create a [subscription](https://console.cloud.google.com/cloudpubsub/subscription/detail/air-report-queue-dl-sub?inv=1&invt=Ab2UTg&project=snap-air&tab=details) with 120s ack deadline 
  2. publish 4k Unacked messages
  3. Create a subscriber with the following setting:         
              100 max outstanding element count 
              8 parallel pull
              ThreadPoolExecutor size of 150   
              MaxAckExtensionPeriod of 3 minutes
              Sleep 30 seconds upon receiving message
  4. Run it for a few minutes
       

#### Code example

```kotlin
        Subscriber.newBuilder(ProjectSubscriptionName.of(PROJECT_ID, "air-report-queue-dl-sub")) { message: PubsubMessage, consumer: AckReplyConsumer ->
            Thread.sleep(30000)
            consumer.ack()
        }.apply {
            setParallelPullCount(8)
            setExecutorProvider(FixedExecutorProvider.create(
                ScheduledThreadPoolExecutor(
                    150
                )
            ))
            setMaxAckExtensionPeriod(Duration.ofMinutes(3))
            setFlowControlSettings(
                FlowControlSettings.newBuilder()
                    .setMaxOutstandingElementCount(100)
                    .build()
            )
        }.build().apply { startAsync().awaitRunning() }
```
Stacktrace: 

```
Jul 09, 2025 1:07:17 PM com.google.cloud.pubsub.v1.StreamingSubscriberConnection$2 onFailure
WARNING: failed to send operations
com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Some acknowledgement ids in the request were invalid. This could be because the acknowledgement ids have expired or the acknowledgement ids were malformed.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:92)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:98)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:84)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1307)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1070)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:819)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:651)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:621)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:569)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.GrpcLoggingInterceptor$1$1.onClose(GrpcLoggingInterceptor.java:98)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:564)
	at io.grpc.internal.ClientCallImpl.access$100(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:729)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:710)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Some acknowledgement ids in the request were invalid. This could be because the acknowledgement ids have expired or the acknowledgement ids were malformed.
	at io.grpc.Status.asRuntimeException(Status.java:532)
	... 21 more
```

I was logging the number of message we're processing in parallel. Initially I see the subscriber was processing 100 requests.
```
Outstanding messages: 0
Outstanding messages: 36
Outstanding messages: 100
Outstanding messages: 100
Outstanding messages: 100
Outstanding messages: 73
Outstanding messages: 19
Outstanding messages: 0
Outstanding messages: 0
Outstanding messages: 0
``` 
But soon as the above error accumulates, its throughput drops to zero in 5 minutes.
My guess is messages are stored in client buffer before they're processed. So they exceeds the 2 minute acknowledge deadline even though the client only spend 30s processing them. And as soon as client failed to ack more than 100 messages, it thinks itself reaches the MaxOutstandingElementCount and stops pulling.
I tried to set this param`setMinDurationPerAckExtensionDuration(Duration.ofMinutes(2))` as suggested in [this](https://cloud.google.com/pubsub/docs/exactly-once-delivery#exactly-once_delivery_support_in_client_libraries) doc. But no luck, I still get the same error

--- a/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java
+++ b/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java
@@ -61,6 +61,8 @@ import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -97,6 +99,7 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
   private final String subscription;
   private final SubscriptionName subscriptionNameObject;
   private final ScheduledExecutorService systemExecutor;
+  private final ExecutorService eodAckCallbackExecutor;
   private final MessageDispatcher messageDispatcher;
 
   private final FlowControlSettings flowControlSettings;
@@ -128,6 +131,7 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
     subscription = builder.subscription;
     subscriptionNameObject = SubscriptionName.parse(builder.subscription);
     systemExecutor = builder.systemExecutor;
+    eodAckCallbackExecutor = builder.eodAckCallbackExecutor;
 
     // We need to set the default stream ack deadline on the initial request, this will be
     // updated by modack requests in the message dispatcher
@@ -455,7 +459,7 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
                       .setSubscription(subscription)
                       .addAllAckIds(ackIdsInRequest)
                       .build());
-      ApiFutures.addCallback(ackFuture, callback, directExecutor());
+      ApiFutures.addCallback(ackFuture, callback, getCallbackExecutor());
       pendingOperations++;
     }
     ackOperationsWaiter.incrementPendingCount(pendingOperations);
@@ -504,7 +508,7 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
                         .addAllAckIds(ackIdsInRequest)
                         .setAckDeadlineSeconds(modackRequestData.getDeadlineExtensionSeconds())
                         .build());
-        ApiFutures.addCallback(modackFuture, callback, directExecutor());
+        ApiFutures.addCallback(modackFuture, callback, getCallbackExecutor());
         pendingOperations++;
       }
     }
@@ -716,6 +720,13 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
     };
   }
 
+  private Executor getCallbackExecutor() {
+    if (!getExactlyOnceDeliveryEnabled()) {
+      return directExecutor();
+    }
+    return eodAckCallbackExecutor;
+  }
+
   /** Builder of {@link StreamingSubscriberConnection StreamingSubscriberConnections}. */
   public static final class Builder {
     private MessageReceiver receiver;
@@ -736,6 +747,7 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
     private boolean useLegacyFlowControl;
     private ScheduledExecutorService executor;
     private ScheduledExecutorService systemExecutor;
+    private ExecutorService eodAckCallbackExecutor;
     private ApiClock clock;
 
     private boolean enableOpenTelemetryTracing;
@@ -823,6 +835,11 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
 
     public Builder setSystemExecutor(ScheduledExecutorService systemExecutor) {
       this.systemExecutor = systemExecutor;
+      return this;
+    }
+
+    public Builder setEodAckCallbackExecutor(ExecutorService eodAckCallbackExecutor) {
+      this.eodAckCallbackExecutor = eodAckCallbackExecutor;
       return this;
     }
 

--- a/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Subscriber.java
+++ b/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Subscriber.java
@@ -51,6 +51,7 @@ import io.opentelemetry.api.trace.Tracer;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadFactory;
@@ -150,6 +151,9 @@ public class Subscriber extends AbstractApiService implements SubscriberInterfac
   // An instantiation of the SystemExecutorProvider used for processing acks
   // and other system actions.
   @Nullable private final ScheduledExecutorService alarmsExecutor;
+  // An executor used for handling ack and modack callbacks when exactly-once delivery is enabled.
+  private final ExecutorService eodAckCallbackExecutor;
+
   private final Distribution ackLatencyDistribution =
       new Distribution(Math.toIntExact(MAX_STREAM_ACK_DEADLINE.getSeconds()) + 1);
 
@@ -199,6 +203,15 @@ public class Subscriber extends AbstractApiService implements SubscriberInterfac
     if (systemExecutorProvider.shouldAutoClose()) {
       backgroundResources.add(new ExecutorAsBackgroundResource((alarmsExecutor)));
     }
+
+    // A cached thread pool will create new threads as needed but can reuse previously constructed
+    // threads when available, which helps to improve performance.
+    ThreadFactory eodAckCallbackThreadFactory =
+        new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("Subscriber-EOD-CallbackExecutor-%d")
+            .build();
+    eodAckCallbackExecutor = Executors.newCachedThreadPool(eodAckCallbackThreadFactory);
 
     TransportChannelProvider channelProvider = builder.channelProvider;
     if (channelProvider.acceptsPoolSize()) {
@@ -416,6 +429,7 @@ public class Subscriber extends AbstractApiService implements SubscriberInterfac
                 .setUseLegacyFlowControl(useLegacyFlowControl)
                 .setExecutor(executor)
                 .setSystemExecutor(alarmsExecutor)
+                .setEodAckCallbackExecutor(eodAckCallbackExecutor)
                 .setClock(clock)
                 .setEnableOpenTelemetryTracing(enableOpenTelemetryTracing)
                 .setTracer(tracer)

--- a/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnectionTest.java
+++ b/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnectionTest.java
@@ -587,6 +587,7 @@ public class StreamingSubscriberConnectionTest {
         .setFlowController(mock(FlowController.class))
         .setExecutor(executor)
         .setSystemExecutor(systemExecutor)
+        .setEodAckCallbackExecutor(systemExecutor)
         .setClock(clock)
         .setMinDurationPerAckExtension(Subscriber.DEFAULT_MIN_ACK_DEADLINE_EXTENSION)
         .setMinDurationPerAckExtensionDefaultUsed(true)

